scripts/o3de/o3de/cmake.py

```diff
@@ -243,3 +242,126 @@ def get_enabled_gem_cmake_file(project_name: str = None,
+
+
+def resolve_gem_dependency_paths(
+        engine_path:pathlib.Path,
+        project_path:pathlib.Path,
+        resolved_gem_dependencies_output_path:pathlib.Path or None):
+    """
+    Resolves gem dependencies for the given engine and project and
+    writes the output to the path provided.  This is used during CMake
+    configuration because writing a CMake depencency resolver would be
+    difficult and Python already has a solver with unit tests.
+    :param engine_path: optional path to the engine, if not provided, the project's engine will be determined 
+    :param project_path: optional path to the project, if not provided the engine path must be provided
+    :param resolved_gem_dependencies_output_path: optional path to a file that will be written 
+        containing a CMake list of gem names and paths.  If not provided, the list is written to STDOUT.
+    :return: 0 for success or non 0 failure code
+    """
+
+    if not engine_path and not project_path:
+        logger.error(f'project path or engine path are required to resolve dependencies')
+        return 1
+
+    if not engine_path:
+        engine_path = manifest.get_project_engine_path(project_path=project_path)
+        if not engine_path:
+            engine_path = manifest.get_this_engine_path()
+            if not engine_path:
+                logger.error('Failed to find a valid engine path for the project at '
+                             f'"{project_path}" which is required to resolve gem dependencies.')
+                return 1
+
+            logger.warning('Failed to determine the correct engine for the project at '
+                           f'"{project_path}", falling back to this engine at {engine_path}.')
+    
+    engine_json_data = manifest.get_engine_json_data(engine_path=engine_path)
+    if not engine_json_data:
+        logger.error('Failed to retrieve engine json data for the engine at '
+                     f'"{engine_path}" which is required to resolve gem dependencies.')
+        return 1
+
+    if project_path:
+        project_json_data = manifest.get_project_json_data(project_path=project_path)
+        active_gem_names = project_json_data.get('gem_names',[])
+        enabled_gems_file = cmake.get_enabled_gem_cmake_file(project_path=project_path)
+        active_gem_names.extend(cmake.get_enabled_gems(enabled_gems_file))
+    else:
+        active_gem_names = engine_json_data.get('gem_names',[])
+
+    # some gem name entries will be dictionaries - convert to a set of strings 
+    gem_names_with_optional_gems = utils.get_gem_names_set(active_gem_names, include_optional=True)
+    if not gem_names_with_optional_gems:
+        logger.info(f'No gem names were found to use as input to resolve gem dependencies.')
+        if resolved_gem_dependencies_output_path:
+            with resolved_gem_dependencies_output_path.open('w') as output:
+                output.write('')
```

Comment:
We could use a function that wraps a [context manager](https://docs.python.org/3/library/contextlib.html?highlight=contextlib#contextlib.contextmanager) to encapsulate the output file
```python
@contextlib.contextmanager
def open_output(output_path:pathlib.Path):   
    output = resolved_gem_dependencies_output_path.open('w') if resolved_gem_dependencies_output_path else sys.stdout
    yield output
    if resolved_gem_dependencies_output_path:
        output.close()

# ...
with open_output(resolved_gem_dependencies_output_path) as output:
    output.write('')
```